### PR TITLE
Handle deleted branch action triggers

### PR DESCRIPTION
**Pull request by glenn-jocher**
## Summary
- report GitHub workflow_dispatch 422 responses instead of posting stale run links
- explain that deleted PR branches need to be restored or replaced before rerunning the command
- cover the deleted-branch dispatch path in tests

## Tests
- python -m pytest tests/test_dispatch_actions.py -v
- ruff check actions/dispatch_actions.py tests/test_dispatch_actions.py
- ruff format --check actions/dispatch_actions.py tests/test_dispatch_actions.py

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Improves GitHub Actions dispatch handling by detecting workflow start failures, reporting clearer PR comment feedback, and bumping the package version to `0.2.17` 🚀

### 📊 Key Changes
- Added **failure tracking** when dispatching workflow files in `dispatch_actions.py` ⚠️
- Now checks GitHub API responses and records workflows that fail to start instead of assuming all dispatches succeed
- Skips the 60-second wait if **all workflows fail immediately**, avoiding unnecessary delays ⏱️
- Includes failed workflows in the returned results with:
  - workflow name
  - file name
  - workflow URL
  - missing run number
  - error message
- Improved PR comment updates to show:
  - a clearer overall status message
  - ✅ successful workflow triggers
  - ❌ failed workflow triggers with specific error details
- Added helpful retry guidance for common issues, such as a deleted PR branch (`no ref found`) 🛠️
- Bumped package version from `0.2.16` to `0.2.17` 📦

### 🎯 Purpose & Impact
- Makes workflow dispatch behavior **more reliable and transparent** for maintainers and contributors
- Helps users quickly understand **why a workflow did not start**, instead of seeing incomplete or misleading results
- Reduces wasted time by avoiding long waits when dispatches fail right away
- Improves the PR experience with clearer action summaries and next-step guidance 💬
- Supports faster troubleshooting of permission, configuration, or branch-related issues, leading to smoother CI/CD operations ✅

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @glenn-jocher, thank you for submitting a `ultralytics/actions` 🚀 PR! This is an automated message to help with review readiness, and an engineer will assist you soon. Please review the checklist below for smooth integration ✨

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/actions/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/actions` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

File: actions/dispatch_actions.py [hunk 1]
```diff
@@ -82,7 +82,20 @@ def trigger_and_get_workflow_info(
...
+                results.append(
+                    {
+                        "name": file.replace(".yml", "").title(),
+                        "file": file,
+                        "url": f"https://github.com/{repo}/actions/workflows/{file}",
+                        "run_number": None,
+                        "error": response.json().get("message", "workflow dispatch failed"),
+                    }
+                )
+                return results
```
Comment (UltralyticsAssistant) [inline actions/dispatch_actions.py]: ⚠️ **HIGH**: Returning immediately on the first `422` drops any workflows that were already dispatched earlier in this loop. For example, if `ci.yml` starts successfully and `docker.yml` fails, the result set will only contain the failure, and `update_comment()` will report that no workflows were started. Append the failure result and continue or break after preserving previously started workflows.

File: actions/dispatch_actions.py [hunk 2]
```diff
@@ -82,7 +82,20 @@ def trigger_and_get_workflow_info(
     try:
         # Trigger all workflows
         for file in workflow_files:
-            event.post(f"{GITHUB_API_URL}/repos/{repo}/actions/workflows/{file}/dispatches", json={"ref": branch})
+            response = event.post(
+                f"{GITHUB_API_URL}/repos/{repo}/actions/workflows/{file}/dispatches", json={"ref": branch}
+            )
+            if response.status_code == 422:
```
Comment (UltralyticsAssistant) [inline actions/dispatch_actions.py]: 💡 **MEDIUM**: This only treats `422` as a dispatch failure. Other non-`204` responses from the GitHub API (`401`, `403`, `404`, `5xx`, etc.) will still fall through to the success path, sleep for 60 seconds, and then produce a comment implying workflows were triggered when they were not. Since this block now handles dispatch failures explicitly, it should reject any unexpected status code, not just `422`.

File: actions/dispatch_actions.py [hunk 3]
```diff
@@ -122,20 +135,31 @@ def update_comment(event, comment_body: str, command: str, triggered_actions: li
...
 (available commands are `{RUN_ALL_KEYWORD}`, `{RUN_CI_KEYWORD}`, and `{RUN_DOCKER_KEYWORD}`):
 
 """
 
     for action in triggered_actions:
-        run_info = f" run {action['run_number']}" if action["run_number"] else ""
-        summary += f"* ✅ [{action['name']}]({action['url']}): `{action['file']}`{run_info}\n"
+        if error := action.get("error"):
+            summary += (
+                f"* ❌ [{action['name']}]({action['url']}): `{action['file']}` failed because `{error}`. "
```
Comment (UltralyticsAssistant) [inline actions/dispatch_actions.py]: 💡 **MEDIUM**: The deleted-branch recovery guidance is shown for every error result, but the only thing we know here is that dispatch failed. GitHub can return `422` for other reasons too, such as a workflow without a `workflow_dispatch` trigger or a disabled workflow. In those cases this comment sends contributors in the wrong direction. Consider showing the restore-branch guidance only when the error message indicates a missing ref/branch.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Good improvement overall: the PR adds much clearer reporting for deleted-branch dispatch failures and extends test coverage for the new behavior. I found a few edge cases in the new failure path that can still misreport workflow state or give misleading guidance.

💬 Posted 3 inline comments

**Comment by UltralyticsAssistant** [comment]
Merged — fantastic improvement, @glenn-jocher! 🚀

“Success depends upon previous preparation, and without such preparation there is sure to be failure.” — Confucius

This update reflects that perfectly: better dispatch failure detection, clearer PR feedback, smarter retry guidance, and less wasted wait time all make Ultralytics Actions more reliable and transparent for everyone. The `0.2.17` bump is a great step toward a smoother contributor and maintainer experience.

Thanks for continuing to strengthen the workflow experience with thoughtful, high-impact improvements.
